src/distributions/uniform.rs

```diff
@@ -534,59 +533,100 @@ macro_rules! uniform_float_impl {
 
             fn new(low: Self::X, high: Self::X) -> Self {
                 assert!(low < high, "Uniform::new called with `low >= high`");
-                let scale = high - low;
-                let offset = low - scale;
-                UniformFloat {
-                    scale: scale,
-                    offset: offset,
+                assert!(low.is_finite() && high.is_finite(), "Uniform::new called with non-finite boundaries");
+
+                let max_rand = (::core::$uty::MAX >> $bits_to_discard)
+                               .into_float_with_exponent(0) - 1.0;
+
+                let mut scale = high - low;
+
+                while scale * max_rand + low >= high {
+                    scale = <$ty>::from_bits(scale.to_bits() - 1);
                 }
+
+                debug_assert!(scale >= 0.0);
+
+                UniformFloat { low, scale }
             }
 
             fn new_inclusive(low: Self::X, high: Self::X) -> Self {
-                assert!(low <= high,
-                        "Uniform::new_inclusive called with `low > high`");
-                let scale = high - low;
-                let offset = low - scale;
-                UniformFloat {
-                    scale: scale,
-                    offset: offset,
+                assert!(low <= high, "Uniform::new_inclusive called with `low > high`");
+                assert!(low.is_finite() && high.is_finite(), "Uniform::new_inclusive called with non-finite boundaries");
```

Comment:
You can reduce the bounds checks a bit: `(NEG_INFINITY < low) & (low <= high) & (high < INFINITY)`.

Reply:
Oh, that's a great idea! But I don't understand why we wouldn't want to short-circuit?

Reply:
Because it might prevent the processor testing multiple things in parallel, and because the common case is that all these tests pass. I don't know; it might be optimisable to the same thing anyway.

Reply:
So it turns out that the old approach is actually slightly faster:

```
&
test gen_range_f32               ... bench:       4,459 ns/iter (+/- 169) = 897 MB/s
test gen_range_f64               ... bench:       6,027 ns/iter (+/- 334) = 1327 MB/s

&&
test gen_range_f32               ... bench:       4,464 ns/iter (+/- 163) = 896 MB/s
test gen_range_f64               ... bench:       6,038 ns/iter (+/- 592) = 1324 MB/s

old
test gen_range_f32               ... bench:       4,382 ns/iter (+/- 135) = 912 MB/s
test gen_range_f64               ... bench:       5,665 ns/iter (+/- 682) = 1412 MB/s
```

Technically none of these differences are stasticially significant, however it definitely looks like the last is faster for f64. And it does provide better error messages.

My best guess is that the old approach ends up being implementable by using bit logic, whereas the `NEG_INF < low` checks has to be done using slower floating point instructions.